shim/src/scopeHandler.js

```diff
@@ -1,128 +1,135 @@
-// the ScopeHandler manages a Proxy which serves as the global scope for the
-// safeEvaluator operation (the Proxy is the argument of a 'with' binding).
-// As described in createSafeEvaluator(), it has several functions:
-// * allow the very first (and only the very first) use of 'eval' to map to
-//   the real (unsafe) eval function, so it acts as a 'direct eval' and can
-//   access its lexical scope (which maps to the 'with' binding, which the
-//   ScopeHandler also controls)
-// * ensure that all subsequent uses of 'eval' map to the safeEvaluator,
-//   which lives as the 'eval' property of the safeGlobal
-// * route all other property lookups at the safeGlobal
-// * hide the unsafeGlobal which lives on the scope chain above the 'with'
-// * ensure the Proxy invariants despite some global properties being frozen
-
-import { getPrototypeOf, objectHasOwnProperty } from './commons';
-
-export class ScopeHandler {
-  // Properties stored on the handler are not available from the proxy.
-
-  // the Proxy is only used by with(), so the Handler only needs to implement
-  // a few properties: has, get, set (which we leave at the default)
-
-  // todo: throw if any traps other than get/set/has are run (e.g.
-  // getOwnPropertyDescriptors, apply, getPrototypeOf) . Make this handler
-  // inherit from a second one whose 'get' property always throws.
-
-  constructor(unsafeRec) {
-    this.unsafeGlobal = unsafeRec.unsafeGlobal;
-    this.unsafeEval = unsafeRec.unsafeEval;
-
-    // this flag allow us to determine if the eval() call is a controlled
+import { assign, create, freeze, getPrototypeOf, objectHasOwnProperty } from './commons';
+import { throwTantrum } from './utilities';
+
+/**
+ * alwayThrowHandler is a proxy handler which throws on any trap called.
+ * It's made from a proxy with a get trap that throws. Its target is
+ * an immutable (frozen) object and is safe to share.
+ */
+const alwayThrowHandler = new Proxy(freeze({}), {
+  get(target, prop) {
+    throwTantrum(`unexpected scope handler trap called: ${prop}`);
+  }
+});
+
+/**
+ * ScopeHandler manages a Proxy which serves as the global scope for the
+ * safeEvaluator operation (the Proxy is the argument of a 'with' binding).
+ * As described in createSafeEvaluator(), it has several functions:
+ * - allow the very first (and only the very first) use of 'eval' to map to
+ *   the real (unsafe) eval function, so it acts as a 'direct eval' and can
+ *    access its lexical scope (which maps to the 'with' binding, which the
+ *   ScopeHandler also controls).
+ * - ensure that all subsequent uses of 'eval' map to the safeEvaluator,
+ *   which lives as the 'eval' property of the safeGlobal.
+ * - route all other property lookups at the safeGlobal.
+ * - hide the unsafeGlobal which lives on the scope chain above the 'with'.
+ * - ensure the Proxy invariants despite some global properties being frozen.
+ */
+export function createScopeHandler(unsafeRec) {
```

Comment:
Very cool. Once again we find that plain objects, and the objects-as-closure pattern, is just better than classes. However, we should follow through and make the could-be-private state variables of the handler into closed over state. Since the handler isn't actually exposed much, this doesn't matter much. But it is safer, cleaner, less verbose, and probably a tad bit faster.

The one state variable that I know of that shouldn't be private is `useUnsafeEvaluator`. But even for it, given how dangerous it is, we should probably make the variable private and expose only a method for setting it to true, to make clear here that this is the only way code outside of here can affect it.

We can start by modifying the function start to

```js
export function createScopeHandler({unsafeGlobal, unsafeEval}) {
  let useUnsafeEvaluator = false;
  ...
```

By switching to the objects as closure pattern, I think all the occurrences of `this.` below can be removed. Avoiding `this` is on the road to enlightenment.